src/routes/files.ts

```diff
@@ -1,14 +1,351 @@
+      )
+
+      if (success) {
+        break
+      } else {
+        requestTimestamp = Date.now()
+      }
+    } catch (err) {
+      next(err)
+    }
+  }
+
+  // Response:
+
+  const responseData: FilesPostResponseData = {
+    paths: fileKeys.reduce((paths, fileKey) => {
+      paths[fileKey] = requestTimestamp
+      return paths
+    }, {})
+  }
+
+  const response: ApiResponse<FilesPostResponseData> = {
+    success: true,
+    data: responseData
+  }
+  res.status(200).json(response)
+})
+
+async function filesUpdateRoutine(
+  body: FilesPostBody,
+  fileKeys: string[],
+  fileKeyInfoMap: FileKeyInfoMap,
+  requestTimestamp: number
+): Promise<boolean> {
+  // Prepare Files Documents:
+
+  const fileRevsResult = await dataStore.fetch({ keys: fileKeys })
+  const storeFileDocuments: StoreFileDocument[] = []
+  const directoryModifications: Map<string, StoreDirectory> = new Map()
+  const repoModifications: Map<
+    string,
+    Pick<StoreRootDocument, 'paths' | 'deleted' | 'timestamp'>
+  > = new Map()
+
+  // Prepare file documents:
+  fileRevsResult.rows.forEach(row => {
+    const fileKey = row.key
+    const storeFile: StoreFile = body.paths[fileKey] || { text: '' }
+    const isDeletion = body.paths[fileKey] === null
+
+    if (!isDeletion) {
+      if ('doc' in row) {
+        let existingFile: StoreFileDocument
+
+        try {
+          existingFile = asStoreFileDocument(row.doc)
+        } catch (err) {
+          const apiClientError: ApiClientError = {
+            status: 422,
+            message:
+              `Unable to write file '${fileKey}'. ` +
+              `Existing document is not a file.`
+          }
+          throw apiClientError
+        }
+
+        // Document will be overwritten
+        storeFileDocuments.push({
+          ...storeFile,
+          _id: fileKey,
+          _rev: row.value.rev
+        })
+      } else {
+        // Document will be inserted
+        storeFileDocuments.push({ ...storeFile, _id: fileKey })
+      }
+    }
+
+    // Directories and Repo Modifications:
+
+    const fileKeyInfo = fileKeyInfoMap.get(fileKey)
+
+    if (fileKeyInfo) {
+      const { repoid, path: filePath } = fileKeyInfo
+
+      // Prepare directory modificaitons:
+      const parentPaths = getParentPathsOfPath(filePath)
+
+      parentPaths.forEach((parentPath, index) => {
+        const directoryKey = `${repoid}:${parentPath}`
+        const existingDirectory = directoryModifications.get(directoryKey)
+        const isRelativePathADir = index < parentPaths.length - 1
+        const relativePath = isRelativePathADir
+          ? getFilenameFromPath(parentPaths[index + 1])
+          : getFilenameFromPath(filePath)
+
+        const directoryModification = updateDirectoryFilePointers(
+          existingDirectory,
+          relativePath,
+          requestTimestamp,
+          isDeletion && !isRelativePathADir
+        )
+
+        directoryModifications.set(directoryKey, directoryModification)
+      })
```

Comment:
Invert the loop to psuedo-guarantee the directory modifications are inserted in the order of directory parent to farthest ancestor of the file.